File: core_challenges/08_algorithm_explanation/algorithm_explanation.py
"""
Algorithm Explanation Challenges

This file contains example algorithms that require both implementation and clear explanation.
Each algorithm focuses on different aspects of problem-solving and optimization.
"""

from typing import List, Dict, Optional, Set, Tuple
from collections import defaultdict, deque
import heapq
import time

# Challenge 1: Advanced Sorting
def hybrid_quicksort(arr: List[int], threshold: int = 10) -> List[int]:
    """
    Implement and explain a hybrid sorting algorithm that combines Quicksort and Insertion Sort.
    
    Key points to explain:
    1. Why combine these algorithms?
    2. How to choose the threshold?
    3. What is the time complexity?
    4. How does it handle different input types?
    
    Example usage:
    >>> arr = [64, 34, 25, 12, 22, 11, 90]
    >>> hybrid_quicksort(arr)
    [11, 12, 22, 25, 34, 64, 90]
    """
    def insertion_sort(arr: List[int], left: int, right: int) -> None:
        """Explain how insertion sort works and why it's good for small arrays"""
        pass

    def partition(arr: List[int], left: int, right: int) -> int:
        """Explain the partitioning strategy and its importance"""
        pass

    def quicksort_recursive(arr: List[int], left: int, right: int) -> None:
        """Explain the recursive approach and the threshold usage"""
        pass

    # Implement the main sorting logic here
    pass

# Challenge 2: Graph Algorithms
class GraphPathfinder:
    """
    Implement and explain different pathfinding algorithms in a graph.
    
    Key points to explain:
    1. Different algorithm choices (Dijkstra, A*, etc.)
    2. Time and space complexity
    3. Heuristic functions
    4. Optimization techniques
    """
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, from_node: int, to_node: int, weight: float):
        """Explain the graph representation choice"""
        pass

    def dijkstra(self, start: int, end: int) -> Tuple[List[int], float]:
        """
        Explain Dijkstra's algorithm:
        1. How it works
        2. Why use a priority queue
        3. Time complexity analysis
        4. Optimization possibilities
        """
        pass

    def a_star(self, start: int, end: int, heuristic) -> Tuple[List[int], float]:
        """
        Explain A* algorithm:
        1. How it improves upon Dijkstra
        2. Heuristic function importance
        3. Admissibility and consistency
        4. Real-world applications
        """
        pass

# Challenge 3: Dynamic Programming
def optimize_investment(
    capital: int,
    projects: List[Dict[str, int]],
    max_projects: int
) -> Tuple[List[str], int]:
    """
    Implement and explain a dynamic programming solution for investment optimization.
    
    Problem:
    Given a capital amount and a list of projects with costs and returns,
    find the optimal selection of projects that maximizes return while:
    - Not exceeding the available capital
    - Not selecting more than max_projects
    - Considering project dependencies
    
    Key points to explain:
    1. State definition
    2. Recurrence relation
    3. Base cases
    4. Space optimization
    
    Example usage:
    >>> projects = [
    ...     {"name": "A", "cost": 100, "return": 200},
    ...     {"name": "B", "cost": 150, "return": 250},
    ...     {"name": "C", "cost": 200, "return": 400}
    ... ]
    >>> optimize_investment(300, projects, 2)
    (["A", "C"], 500)
    """
    def create_dp_table() -> List[List[List[int]]]:
        """Explain the 3D DP table structure"""
        pass

    def fill_dp_table() -> None:
        """Explain the bottom-up approach"""
        pass

    def backtrack_solution() -> Tuple[List[str], int]:
        """Explain the solution reconstruction process"""
        pass

    # Implement the main optimization logic here
    pass

# Challenge 4: String Algorithms
def pattern_matching(text: str, pattern: str) -> List[int]:
    """
    Implement and explain an efficient pattern matching algorithm (e.g., KMP, Boyer-Moore).
    
    Key points to explain:
    1. Preprocessing phase
    2. Matching phase
    3. Time complexity analysis
    4. Space-time tradeoffs
    
    Example usage:
    >>> text = "AABAACAADAABAAABAA"
    >>> pattern = "AABA"
    >>> pattern_matching(text, pattern)
    [0, 9, 13]
    """
    def build_pattern_table() -> List[int]:
        """Explain the pattern preprocessing"""
        pass

    def find_matches() -> List[int]:
        """Explain the matching process"""
        pass

    # Implement the pattern matching logic here
    pass 

File: core_challenges/08_algorithm_explanation/algorithm_explanation.js
/**
 * Algorithm Explanation Challenges
 * 
 * This file contains example algorithms that require both implementation and clear explanation.
 * Each algorithm focuses on different aspects of problem-solving and optimization.
 */

// Challenge 1: Advanced Sorting
/**
 * Implements and explains a hybrid sorting algorithm that combines Quicksort and Insertion Sort.
 * 
 * Key points to explain:
 * 1. Why combine these algorithms?
 * 2. How to choose the threshold?
 * 3. What is the time complexity?
 * 4. How does it handle different input types?
 * 
 * @param {Array<number>} arr - Array to sort
 * @param {number} threshold - Threshold for switching to insertion sort
 * @returns {Array<number>} Sorted array
 * 
 * @example
 * const arr = [64, 34, 25, 12, 22, 11, 90];
 * hybridQuicksort(arr);
 * // Returns [11, 12, 22, 25, 34, 64, 90]
 */
function hybridQuicksort(arr, threshold = 10) {
    /**
     * Explain how insertion sort works and why it's good for small arrays
     * @param {Array<number>} arr - Array to sort
     * @param {number} left - Left boundary
     * @param {number} right - Right boundary
     */
    function insertionSort(arr, left, right) {
        // Implement insertion sort
    }

    /**
     * Explain the partitioning strategy and its importance
     * @param {Array<number>} arr - Array to partition
     * @param {number} left - Left boundary
     * @param {number} right - Right boundary
     * @returns {number} Partition index
     */
    function partition(arr, left, right) {
        // Implement partition
    }

    /**
     * Explain the recursive approach and the threshold usage
     * @param {Array<number>} arr - Array to sort
     * @param {number} left - Left boundary
     * @param {number} right - Right boundary
     */
    function quicksortRecursive(arr, left, right) {
        // Implement quicksort with threshold
    }

    // Implement the main sorting logic here
}

// Challenge 2: Graph Algorithms
/**
 * Implements and explains different pathfinding algorithms in a graph.
 * 
 * Key points to explain:
 * 1. Different algorithm choices (Dijkstra, A*, etc.)
 * 2. Time and space complexity
 * 3. Heuristic functions
 * 4. Optimization techniques
 */
class GraphPathfinder {
    constructor() {
        this.graph = new Map();
    }

    /**
     * Explain the graph representation choice
     * @param {number} fromNode - Source node
     * @param {number} toNode - Target node
     * @param {number} weight - Edge weight
     */
    addEdge(fromNode, toNode, weight) {
        // Implement edge addition
    }

    /**
     * Explain Dijkstra's algorithm:
     * 1. How it works
     * 2. Why use a priority queue
     * 3. Time complexity analysis
     * 4. Optimization possibilities
     * 
     * @param {number} start - Start node
     * @param {number} end - End node
     * @returns {[Array<number>, number]} Path and total cost
     */
    dijkstra(start, end) {
        // Implement Dijkstra's algorithm
    }

    /**
     * Explain A* algorithm:
     * 1. How it improves upon Dijkstra
     * 2. Heuristic function importance
     * 3. Admissibility and consistency
     * 4. Real-world applications
     * 
     * @param {number} start - Start node
     * @param {number} end - End node
     * @param {Function} heuristic - Heuristic function
     * @returns {[Array<number>, number]} Path and total cost
     */
    aStar(start, end, heuristic) {
        // Implement A* algorithm
    }
}

// Challenge 3: Dynamic Programming
/**
 * Implements and explains a dynamic programming solution for investment optimization.
 * 
 * Problem:
 * Given a capital amount and a list of projects with costs and returns,
 * find the optimal selection of projects that maximizes return while:
 * - Not exceeding the available capital
 * - Not selecting more than max_projects
 * - Considering project dependencies
 * 
 * Key points to explain:
 * 1. State definition
 * 2. Recurrence relation
 * 3. Base cases
 * 4. Space optimization
 * 
 * @param {number} capital - Available capital
 * @param {Array<Object>} projects - List of projects
 * @param {number} maxProjects - Maximum number of projects
 * @returns {[Array<string>, number]} Selected projects and total return
 * 
 * @example
 * const projects = [
 *   { name: "A", cost: 100, return: 200 },
 *   { name: "B", cost: 150, return: 250 },
 *   { name: "C", cost: 200, return: 400 }
 * ];
 * optimizeInvestment(300, projects, 2);
 * // Returns [["A", "C"], 500]
 */
function optimizeInvestment(capital, projects, maxProjects) {
    /**
     * Explain the 3D DP table structure
     * @returns {Array<Array<Array<number>>>} DP table
     */
    function createDpTable() {
        // Implement DP table creation
    }

    /**
     * Explain the bottom-up approach
     */
    function fillDpTable() {
        // Implement DP table filling
    }

    /**
     * Explain the solution reconstruction process
     * @returns {[Array<string>, number]} Solution
     */
    function backtrackSolution() {
        // Implement solution backtracking
    }

    // Implement the main optimization logic here
}

// Challenge 4: String Algorithms
/**
 * Implements and explains an efficient pattern matching algorithm (e.g., KMP, Boyer-Moore).
 * 
 * Key points to explain:
 * 1. Preprocessing phase
 * 2. Matching phase
 * 3. Time complexity analysis
 * 4. Space-time tradeoffs
 * 
 * @param {string} text - Text to search in
 * @param {string} pattern - Pattern to search for
 * @returns {Array<number>} Array of match positions
 * 
 * @example
 * const text = "AABAACAADAABAAABAA";
 * const pattern = "AABA";
 * patternMatching(text, pattern);
 * // Returns [0, 9, 13]
 */
function patternMatching(text, pattern) {
    /**
     * Explain the pattern preprocessing
     * @returns {Array<number>} Pattern table
     */
    function buildPatternTable() {
        // Implement pattern table building
    }

    /**
     * Explain the matching process
     * @returns {Array<number>} Match positions
     */
    function findMatches() {
        // Implement pattern matching
    }

    // Implement the pattern matching logic here
} 